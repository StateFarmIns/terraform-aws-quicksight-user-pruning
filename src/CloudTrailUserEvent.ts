import { Event } from '@aws-sdk/client-cloudtrail'

export class CloudTrailUserEvent {
	// arn:aws:sts::account:assumed-role/ROLE_NAME/SESSION_NAME
	iamRole: string 
	stsSession: string 
	eventTime: Date

	constructor(event: Event) {
		const cloudTrailEvent = JSON.parse(event.CloudTrailEvent)
		// TODO: What about IAM user logins?
		const [role, session] =  cloudTrailEvent.userIdentity.arn.split('/').slice(1) // arn:aws:sts::account:assumed-role/ROLE_NAME/SESSION_NAME
		
		this.iamRole = role
		this.stsSession = session
		this.eventTime = event.EventTime
	}
}