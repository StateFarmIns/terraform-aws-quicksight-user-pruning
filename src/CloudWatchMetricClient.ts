import { CloudWatchClient, MetricDatum, PutMetricDataCommand } from '@aws-sdk/client-cloudwatch'

export class CloudWatchMetricClient {
	private metricData: MetricDatum[] = []
	private cloudWatchClient = new CloudWatchClient({})

	// Normally we'd worry about max metric data per request but the max is currently 20 and we won't be publishing that many metrics
	public queueMetric(metricDatum: MetricDatum) {
		this.metricData.push(metricDatum)
	}

	public async emitQueuedMetrics() {
		const putMetricDataCommand = new PutMetricDataCommand({
			Namespace: process.env.AWS_LAMBDA_FUNCTION_NAME,
			MetricData: this.metricData,
		})

		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Emitting metrics: ${JSON.stringify(putMetricDataCommand)}`)

		await this.cloudWatchClient.send(putMetricDataCommand)

		this.metricData = []
	}
}