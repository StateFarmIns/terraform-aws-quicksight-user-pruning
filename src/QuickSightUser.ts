import { User } from '@aws-sdk/client-quicksight'

export class QuickSightUser {
	arn: string // arn:aws:quicksight:<REGION>:<ACCOUNT>:user/default/<IAM ROLE NAME>/<STS SESSION>
	email: string
	lastAccess: Date // Comes from CloudTrail
	role: QuickSightRole
	username: string // <IAM ROLE>/<STS SESSION>
	iamRole: string
	stsSession: string
	invalid: boolean // User is invalid if username is "N/A". See README.md for more details

	constructor(quickSightUser: User) {
		this.arn = quickSightUser.Arn
		this.email = quickSightUser.Email
		this.role = QuickSightRole[quickSightUser.Role as keyof typeof QuickSightRole] // Only values that will be returned // also the keyof typeof hack came from here https://stackoverflow.com/a/42623905
		this.username = quickSightUser.UserName
		this.invalid = this.username === 'N/A'
		const [iamRole, stsSession] = this.username.split('/')
		this.iamRole = iamRole
		this.stsSession = stsSession
	}
}

export enum QuickSightRole {
	READER = 'READER', AUTHOR = 'AUTHOR', ADMIN = 'ADMIN'
}