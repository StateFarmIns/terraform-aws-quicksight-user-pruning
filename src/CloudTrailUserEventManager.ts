import { CloudTrailClient, LookupEventsCommand } from '@aws-sdk/client-cloudtrail'
import { CloudTrailUserEvent } from './CloudTrailUserEvent'
import { QuickSightUser } from './QuickSightUser'

export class CloudTrailUserEventManager {
	private cloudTrailClient = new CloudTrailClient({})

	// Filters out events that do not originate with someone's e-mail address
	// such as any automated process which touches QuickSight, including this one
	public async retrieveQuickSightUserEvents(startDate: Date): Promise<CloudTrailUserEvent[]> {
		const events: CloudTrailUserEvent[] = []

		let nextToken: string = null // This is annoying; null doesn't work in the QuickSight API but empty string doesn't work in the CloudTrail API
		do {
			const lookupEventsCommand = new LookupEventsCommand({
				LookupAttributes: [
					{
						AttributeKey: 'EventSource',
						AttributeValue: 'quicksight.amazonaws.com',
					},
				],
				StartTime: startDate,
				NextToken: nextToken,
			})

			const lookupEventsResult = await this.cloudTrailClient.send(lookupEventsCommand)

			nextToken = lookupEventsResult.NextToken

			// Filter out users who do not have an @ sign in the name, this limiting events to only
			// those that were user-initiated (e-mail address)
			const filteredEvents = lookupEventsResult.Events.filter((event) => event.Username.includes('@'))

			for (const event of filteredEvents) {
				events.push(new CloudTrailUserEvent(event))
			}
		} while (nextToken)

		return events
	}

	public getLastAccessDate(user: QuickSightUser, events: CloudTrailUserEvent[]): Date {
		const thisParticularUserEvents = events.filter((event) => event.email === user.email && event.iamRoleId === user.iamRoleId)

		if (thisParticularUserEvents.length === 0) {
			return new Date(0) // Return 1970 as last access date to make logic easier down the road
		}

		// Sort newest to oldest
		return thisParticularUserEvents.sort((a, b) => a.eventTime <= b.eventTime ? 1 : -1)[0].eventTime
	}
}