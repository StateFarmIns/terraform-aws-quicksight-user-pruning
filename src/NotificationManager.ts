import { SendEmailCommand, SESClient } from '@aws-sdk/client-ses'
import { QuickSightUser } from './QuickSightUser'

export class NotificationManager {
	private sesClient = new SESClient({})
	private noLastAccessDate = new Date(0).toLocaleString()

	public async notifyUser(user: QuickSightUser) {
		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Notifying ${JSON.stringify(user, null, 1)}`)
		
		// If the date is 1970, we didn't find any activity for the user
		const lastAccess =
			user.lastAccess.toLocaleString() === this.noLastAccessDate
				? `not in the last ${process.env.deleteDays} days`
				: user.lastAccess.toLocaleDateString()

		const subject = `I: QuickSight user ${user.username} in account ${process.env.accountAlias} will be deleted`
		const body = `In order to <a href="https://aws.amazon.com/quicksight/pricing/">control costs ($24/month/user)</a>,
    your AWS QuickSight user '${user.username}' 
    will be deleted when you have not accessed QuickSight in ${process.env.deleteDays} days. 
    CloudTrail records indicate that your last access was ${lastAccess}.<br /><br />
    
    You can keep your user by logging in and taking actions in QuickSight. 
    Alternatively, you can do nothing, and your user will be automatically deleted. 
    Never fear, you simply log back into QuickSight and enter your e-mail, and your user gets recreated. 
    No dashboards or other resources will be deleted at this time (we may clean those up in the future, with notification, of course).<br /><br />
    
    Account Number: ${process.env.awsAccountId}<br />
    Account Alias: ${process.env.accountAlias}<br />
    Assumed Role: ${user.iamRole}<br />
    Email: ${user.email}<br /><br />
  
    For further assistance, contact <a href="mailto:${process.env.contact}">${process.env.contact}</a>. Thank you and have a fantastically fun-filled day ðŸ¥³.
    `

		const sendEmailCommand = new SendEmailCommand({
			Source: `${process.env.accountAlias} Cloud Account Administrator<${process.env.from}>`,
			Destination: {
				ToAddresses: [user.email],
				CcAddresses: JSON.parse(process.env.cc),
			},
			ReplyToAddresses: [process.env.replyTo],
			SourceArn: process.env.sesArn,
			Message: {
				Subject: { Data: subject },
				Body: {
					Html: { Data: body },
				},
			},
		})

		await this.sesClient.send(sendEmailCommand)
	}
}