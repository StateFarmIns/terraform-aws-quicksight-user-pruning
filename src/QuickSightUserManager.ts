import { DeleteUserCommand, ListUsersCommand, QuickSightClient } from '@aws-sdk/client-quicksight'
import { QuickSightUser } from './QuickSightUser'

export class QuickSightUserManager {
	private awsAccountId = process.env.awsAccountId
	private static namespace = 'default'
	private quickSightClient = new QuickSightClient({})
  
	public async retrieveUsers(): Promise<QuickSightUser[]> {
		const quickSightUsers: QuickSightUser[] = []

		let nextToken = '' // originally I tried null, but it causes API signature issues.
		do {
			const listUsersCommand = new ListUsersCommand({ AwsAccountId: this.awsAccountId, Namespace: QuickSightUserManager.namespace, NextToken: nextToken })
			const quickSightUserListResponse = await this.quickSightClient.send(listUsersCommand)

			nextToken = quickSightUserListResponse.NextToken

			quickSightUserListResponse.UserList.forEach((user) => quickSightUsers.push(new QuickSightUser(user)))
		} while (nextToken)

		return quickSightUsers
	}

	public async deleteUser(user: QuickSightUser) {
		// Stryker disable next-line all "I do not care about mutating console statements"
		console.debug(`Deleting ${JSON.stringify(user, null, 1)}`)
		
		const deleteUserCommand = new DeleteUserCommand({
			AwsAccountId: this.awsAccountId,
			Namespace: QuickSightUserManager.namespace,
			UserName: user.username,
		})

		await this.quickSightClient.send(deleteUserCommand)
	}
}