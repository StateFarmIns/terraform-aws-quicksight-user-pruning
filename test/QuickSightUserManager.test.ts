import { DeleteUserCommand, ListUsersCommand, QuickSightClient, User } from '@aws-sdk/client-quicksight'
import { mockClient } from 'aws-sdk-client-mock'
import { QuickSightUser } from '../src/QuickSightUser'
import { QuickSightUserManager } from '../src/QuickSightUserManager'

process.env.awsAccountId = '1234567890'

const quickSightUserManager: QuickSightUserManager = new QuickSightUserManager()
const quickSightMock = mockClient(QuickSightClient)
const validUsers: User[] = [ // User is a type from AWS SDK client
	{
		Active: true,
		Arn: 'arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-admin-role/john.smith@example.com',
		Email: 'john.smith@example.com',
		PrincipalId: 'federated/iam/ARIAGRGHRGGERHQWOJ:john.smith@example.com',
		UserName: 'quicksight-admin-role/john.smith@example.com',
		Role: 'ADMIN',
	},
	{
		Active: true,
		Arn: 'arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-author-role/jill.hill@example.com',
		Email: 'jill.hill@example.com',
		PrincipalId: 'federated/iam/ARIAGRGHTRHGERHQWOJ:jill.hill@example.com',
		UserName: 'quicksight-author-role/jill.hill@example.com',
		Role: 'AUTHOR',
	},
	{
		Active: true,
		Arn: 'arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-reader/lowly.reader@example.com',
		Email: 'lowly.reader@example.com',
		PrincipalId: 'federated/iam/ARIAGERHIGWFEHQOFIH:lowly.reader@example.com',
		UserName: 'quicksight-reader-role/lowly.reader@example.com',
		Role: 'READER',
	},
]

const validQuickSightUsers: QuickSightUser[] = validUsers.map((user) => new QuickSightUser(user)) // QuickSightUser is our custom type

afterEach(() => {
	quickSightMock.reset()
})

describe('QuickSightUserManager', () => {
	describe('retrieveUsers', () => {
		it('gets a valid one-page list of users', async () => {
			quickSightMock.on(ListUsersCommand).resolves({
				UserList: validUsers,
			})

			const users = await quickSightUserManager.retrieveUsers()

			expect(quickSightMock.calls()).toHaveLength(1)

			const expectedInput = { AwsAccountId: process.env.awsAccountId, Namespace: 'default', NextToken: '' }
			expect(quickSightMock.call(0).args[0].input).toStrictEqual(expectedInput)

			expect(users).toMatchSnapshot()
		})

		it('gets a valid two-page list of users', async () => {
			quickSightMock.on(ListUsersCommand).resolvesOnce({
				UserList: [validUsers[0]],
				NextToken: 'GetMoreResults',
			}).resolvesOnce({
				UserList: [validUsers[1]],
			})


			const users = await quickSightUserManager.retrieveUsers()

			expect(quickSightMock.calls()).toHaveLength(2)

			const expectedInput = { AwsAccountId: process.env.awsAccountId, Namespace: 'default', NextToken: '' }
			expect(quickSightMock.call(0).args[0].input).toStrictEqual(expectedInput)

			const secondExpectedInput = { ...expectedInput, NextToken: 'GetMoreResults' }
			expect(quickSightMock.call(1).args[0].input).toStrictEqual(secondExpectedInput)

			expect(users).toMatchSnapshot()
		})

		it('fails when receiving exception', async () => {
			const error: Error = { name: 'BadError', message: 'Oh no!' }
			quickSightMock.on(ListUsersCommand).rejects(error)

			await expect(quickSightUserManager.retrieveUsers()).rejects.toThrowError(error)
		})
	})

	describe('deleteUser', () => {
		it('successfully deletes a user', async () => {
			quickSightMock.on(DeleteUserCommand).resolves({})

			await quickSightUserManager.deleteUser(validQuickSightUsers[0])

			expect(quickSightMock.calls()).toHaveLength(1)

			const expectedInput = { AwsAccountId: process.env.awsAccountId, Namespace: 'default', UserName: validQuickSightUsers[0].username }
			expect(quickSightMock.call(0).args[0].input).toStrictEqual(expectedInput)
		})

		it('fails when receiving exception', async () => {
			const error: Error = { name: 'BadError', message: 'Oh no!' }
			quickSightMock.on(DeleteUserCommand).rejects(error)

			await expect(quickSightUserManager.deleteUser(validQuickSightUsers[0])).rejects.toThrowError(error)
		})
	})
})