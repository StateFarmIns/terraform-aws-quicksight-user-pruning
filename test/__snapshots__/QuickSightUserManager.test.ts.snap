// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickSightUserManager retrieveUsers gets a valid one-page list of users 1`] = `
Array [
  QuickSightUser {
    "arn": "arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-admin-role/john.smith@example.com",
    "email": "john.smith@example.com",
    "iamRoleId": "ARIAGRGHRGGERHQWOJ",
    "iamRoleName": "quicksight-admin-role",
    "invalid": false,
    "role": "ADMIN",
    "username": "quicksight-admin-role/john.smith@example.com",
  },
  QuickSightUser {
    "arn": "arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-author-role/jill.hill@example.com",
    "email": "jill.hill@example.com",
    "iamRoleId": "ARIAGRGHTRHGERHQWOJ",
    "iamRoleName": "quicksight-author-role",
    "invalid": false,
    "role": "AUTHOR",
    "username": "quicksight-author-role/jill.hill@example.com",
  },
  QuickSightUser {
    "arn": "arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-reader/lowly.reader@example.com",
    "email": "lowly.reader@example.com",
    "iamRoleId": "ARIAGERHIGWFEHQOFIH",
    "iamRoleName": "quicksight-reader-role",
    "invalid": false,
    "role": "READER",
    "username": "quicksight-reader-role/lowly.reader@example.com",
  },
]
`;

exports[`QuickSightUserManager retrieveUsers gets a valid two-page list of users 1`] = `
Array [
  QuickSightUser {
    "arn": "arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-admin-role/john.smith@example.com",
    "email": "john.smith@example.com",
    "iamRoleId": "ARIAGRGHRGGERHQWOJ",
    "iamRoleName": "quicksight-admin-role",
    "invalid": false,
    "role": "ADMIN",
    "username": "quicksight-admin-role/john.smith@example.com",
  },
  QuickSightUser {
    "arn": "arn:aws:quicksight:us-east-1:1234567890:user/default/quicksight-author-role/jill.hill@example.com",
    "email": "jill.hill@example.com",
    "iamRoleId": "ARIAGRGHTRHGERHQWOJ",
    "iamRoleName": "quicksight-author-role",
    "invalid": false,
    "role": "AUTHOR",
    "username": "quicksight-author-role/jill.hill@example.com",
  },
]
`;
